internal/storage/prefixdb.go

```diff
@@ -0,0 +1,632 @@
+func (b prefixDBBatch) Write() error {
+	if err := b.source.Write(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch write: %w", err)
+	}
+	return nil
+}
+
+// WriteSync applies the batch to the database. WriteSync guarantees that the batch
+// is persisted to disk before returning.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) WriteSync() error {
+	if err := b.source.WriteSync(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch write: %w", err)
+	}
+	return nil
+}
+
+// Close closes the batch without committing it. Close is idempotent, but calling
+// other methods on the batch after closing it will return an error.
+//
+// It implements the [Batch] interface for type prefixDBBatch.
+func (b prefixDBBatch) Close() error {
+	if err := b.source.Close(); err != nil {
+		return fmt.Errorf("prefixed DB namespace batch close: %w", err)
+	}
+	return nil
+}
+
+// prefixDBIterator wraps an [Iterator] interface representing an iterator iterating
+// over a range of key/value pairs in an underlying database. It will range over
+// keys that begin with the given prefix. That is, prefixDBIterator ensures that all
+// keys that it iterates over are scoped to the provided prefix.
+//
+// It is the caller's responsibility to call [Close] on it when done.
+// prefixDBIterator is not safe for concurrent use, but it is safe to use multiple
+// iterators concurrently. Callers cannot write to the underlying database
+// while there exists an iterator over it.
+// If there is an error during any operation, it is stored in the Iterator and can
+// be retrieved via the [Error] method.
+//
+// It implements the [Iterator] interface.
+type prefixDBIterator struct {
+	source Iterator
+	prefix []byte
+	start  []byte
+	end    []byte
+	valid  bool
+	err    error
+}
+
+// compile-time check: does *prefixDBIterator satisfy the Iterator interface?
+var _ Iterator = (*prefixDBIterator)(nil)
+
+// newPrefixDBIterator returns a new prefixDBIterator to iterate over a range of
+// database key/value pairs that begin with the given prefix.
+func newPrefixDBIterator(
+	prefix, start, end []byte,
+	source Iterator,
+) (*prefixDBIterator, error) { //nolint:unparam
+	itInvalid := &prefixDBIterator{
+		prefix: prefix,
+		start:  start,
+		end:    end,
+		source: source,
+		valid:  false,
+	}
+
+	// Empty keys are not allowed, so if a key exists in the database that exactly
+	// matches the prefix, we need to skip it.
+	if source.Valid() && bytes.Equal(source.Key(), prefix) {
+		// The key is going to be lexicograpically smaller than the first
+		// "correct" key of the form prefix+key, because
+		// prefix < prefix+key, e.g., "foo" < "fooa".
+		// Therefore, we only need to skip it to position the iterator at the first
+		// "correct" key.
+		source.Next()
+	}
+
+	if !source.Valid() || !bytes.HasPrefix(source.Key(), prefix) {
+		return itInvalid, nil
+	}
```

Comment:
Why do we create the `itInvalid` object? Since it's an invalid iterator, any method call on it will cause a panic, thus making it useless. Why not return an error here instead?

Reply:
I think we should return an error